scarb/src/ops/proc_macro_server/connection.rs

```diff
@@ -0,0 +1,98 @@
+use crossbeam_channel::{Receiver, Sender};
+use proc_macro_server_api::{RpcRequest, RpcResponse};
+use std::{
+    io::{BufRead, Write},
+    thread::JoinHandle,
+};
+
+pub struct Connection {
+    pub sender: Sender<RpcResponse>,
+    pub receiver: Receiver<RpcRequest>,
+    io_threads: IoThreads,
+}
+
+impl Connection {
+    pub fn new() -> Self {
+        let (reader_sender, reader_receiver) = crossbeam_channel::bounded(0);
+        let (writer_sender, writer_receiver) = crossbeam_channel::bounded(0);
+
+        let reader = std::thread::spawn(move || {
+            let stdin = std::io::stdin();
+            let mut stdin = stdin.lock();
+
+            let mut line = String::new();
+
+            loop {
+                line.clear();
+
+                if stdin.read_line(&mut line).is_err() {
+                    eprintln!("Error occurred while reading from stdin");
+
+                    break;
+                }
+
+                if line.is_empty() {
+                    continue;
+                }
+
+                let Ok(request) = serde_json::from_str(&line) else {
+                    eprintln!("Error occurred while deserializing request, used input:\n{line}");
+
+                    break;
+                };
+
+                if reader_sender.send(request).is_err() {
+                    eprintln!("Error occurred while sending request to worker threads");
+
+                    break;
+                }
```

Comment:
a lot of vertical noise here imo

```suggestion
                if stdin.read_line(&mut line).is_err() {
                    eprintln!("Error occurred while reading from stdin");
                    break;
                }

                if line.is_empty() {
                    continue;
                }

                let Ok(request) = serde_json::from_str(&line) else {
                    eprintln!("Error occurred while deserializing request, used input:\n{line}");
                    break;
                };

                if reader_sender.send(request).is_err() {
                    eprintln!("Error occurred while sending request to worker threads");
                    break;
                }
```